pkg/providers/vault/factory.go

```diff
@@ -0,0 +1,94 @@
+/*******************************************************************************
+ * Copyright 2019 Dell Inc.
+ * Copyright 2020 Intel Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package vault
+
+import (
+	"context"
+	"time"
+
+	"github.com/edgexfoundry/go-mod-secrets/pkg"
+)
+
+// a map variable to handle the case of the same caller to have
+// multiple secret clients with potentially the same tokens while renewing token
+// in the background go-routine
+type vaultTokenToCancelFuncMap map[string]context.CancelFunc
+
+type secretClientFactory struct {
+	tokenCancelFunc vaultTokenToCancelFuncMap
```

Comment:
Do we need a lock/semaphore to protect concurrent updates to this data structure?

Reply:
there is no concurrent update by internal code here unless the caller is running concurrently.

Reply:
This is a module implementation intended to be used by other services and you can't predict whether or not the consuming services will use it concurrently.  As such, you should ensure that it is thread safe.  Please add appropriate mutex support to ensure map usage is thread safe.

Reply:
ok

Reply:
Added